sycl/include/CL/sycl/INTEL/online_compiler.hpp

```diff
@@ -142,51 +175,48 @@ template <source_language Lang> class online_compiler {
 
   /// Target device stepping (implementation defined)
   std::string DeviceStepping;
-  // }
+
+  /// Handles to helper functions used by the implementation.
+  void *CompileToSPIRVHandle = nullptr;
+  void *FreeSPIRVOutputsHandle = nullptr;
 };
 
 // Specializations of the online_compiler class and 'compile' function for
 // particular languages and parameter types.
 
-/// Compiles given OpenCL source. May throw \c online_compile_error.
-/// @param src - contents of the source
+/// Compiles the given OpenCL source. May throw \c online_compile_error.
+/// @param src - contents of the source.
+/// @param options - compilation options (implementation defined); standard
+///   OpenCL JIT compiler options must be supported.
+template <>
+template <>
+__SYCL_EXPORT std::vector<byte>
+online_compiler<source_language::opencl_c>::compile(
+    const std::string &src, const std::vector<std::string> &options);
+
+/// Compiles the given OpenCL source. May throw \c online_compile_error.
+/// @param src - contents of the source.
 template <>
 template <>
 std::vector<byte>
 online_compiler<source_language::opencl_c>::compile(const std::string &src) {
-  // real implementation will call some non-templated impl function here
-  return std::vector<byte>{};
+  return compile(src, std::vector<std::string>{});
 }
 
-/// Compiles given OpenCL source. May throw \c online_compile_error.
-/// @param src - contents of the source
-/// @param options - compilation options (implementation defined); standard
-///   OpenCL JIT compiler options must be supported
+/// Compiles the given CM source \p src.
+/// @param src - contents of the source.
+/// @param options - compilation options (implementation defined).
 template <>
 template <>
-std::vector<byte> online_compiler<source_language::opencl_c>::compile(
-    const std::string &src, const std::vector<std::string> &options) {
-  // real implementation will call some non-templated impl function here
-  return std::vector<byte>{};
-}
+__SYCL_EXPORT std::vector<byte> online_compiler<source_language::cm>::compile(
+    const std::string &src, const std::vector<std::string> &options);
```

Comment:
I have the feeling that
```suggestion
    const std::string &src, const std::vector<std::string> &options = {});
```
would make the first overload useless.

Reply:
I surely thought about doing that, but it doesn't work and the attempt to add the default argument, then remove the overload accepting only 'src' causes this error during SYCL build phase:

llvm/sycl/include/CL/sycl/INTEL/online_compiler.hpp:195:73: error: default argument specified in explicit specialization [ -fpermissive ]
     const std::string &src, const std::vector< std::string > &options = {});

Reply:
I see, because you have the main definition as
```
template <typename... Tys>
  std::vector<byte> compile(const std::string &src, const Tys &... args);
```
you cannot add this default parameter there either.
Actually I am even surprised that it is possible to specialize a variadic argument by a non variadic one...

But, at the end, why do you need to declare some super generic `compile()` that you do not really use, instead of having extensions just adding incrementally the overload they want?
I have the feeling that this specialization just adds some complexity for nothing valuable.
And the variadic template derail some simple implementations like:
https://godbolt.org/z/Yfo9bd
```
#include <iostream>
#include <string>
#include <vector>
struct a {
    template <typename... Types>
    static void compile(Types... args) {
        std::cout << "the generic default" << std::endl;
    }
    static void compile(std::string arg, std::vector<std::string> args = {}) {
        std::cout << "single arg with vector" << std::endl;
    }
};

int main() {
   // This one is captured by the generic default because a `const char*` is not a `std::string`, but do we need this at all?
   a::compile("default");
   a::compile(std::string{});
   a::compile(std::string{}, {"opt"});
   a::compile("", {"opt"});
   a::compile(std::string{}, {});
   a::compile(std::string{}, {"opt"});
     return 0;
}
```

Reply:
> But, at the end, why do you need to declare some super generic compile() that you do not really use, instead of having extensions just adding incrementally the overload they want?
I have the feeling that this specialization just adds some complexity for nothing valuable.
And the variadic template derail some simple implementations like:
https://godbolt.org/z/Yfo9bd

@kbobrovs , is the one who wrote the feature SPEC, perhaps Konst can answer your question.

Reply:
The assumption here is that different languages may have very different parameter sets for compile. Variadic templates provide the needed freedom. Concrete specializations fix variants used for particular language.

To avoid "derailing" mentioned above, we simply don't provide `compile` specialization pairs which conflict in that way.

Overloading is surely an option too, but some overloads will be unavailable for certain languages, which adds confusion. Isn't variadic template is ok C++ way to go when parameter list can be arbitrary?

Adding @Pennycook and @rolandschulz for possible comments.